tests/common/test_SingletonPerThread.py

```diff
@@ -0,0 +1,144 @@
+import uuid
+from multiprocessing import Queue
+from threading import Thread, currentThread
+
+from azurelinuxagent.common.SingletonPerThread import SingletonPerThread
+from tests.tools import AgentTestCase
+
+
+class TestClassToTestSingletonPerThread(SingletonPerThread):
+
+    def __init__(self):
+        # Set the name of the object to the current thread name
+        self.name = currentThread().getName()
+        # Unique identifier for a class object
+        self.uuid = str(uuid.uuid4())
+
+
+class TestSingletonPerThread(AgentTestCase):
+
+    THREAD_NAME_1 = 'thread-1'
+    THREAD_NAME_2 = 'thread-2'
+
+    def setUp(self):
+        super(TestSingletonPerThread, self).setUp()
+        # In a multi-threaded environment, exceptions thrown in the child thread will not be propagated to the parent
+        # thread. In order to achieve that, adding all exceptions to a Queue and then checking that in parent thread.
+        self.errors = Queue()
+        TestClassToTestSingletonPerThread.clear()
+
+    def _setup_mutithread_and_execute(self, func1, args1, func2, args2):
+
+        t1 = Thread(target=func1, args=args1)
+        t2 = Thread(target=func2, args=args2)
+        t1.setName(self.THREAD_NAME_1)
+        t2.setName(self.THREAD_NAME_2)
+        t1.start()
+        t2.start()
+        t1.join()
+        t2.join()
+
+        errs = []
+        while not self.errors.empty():
+            errs.append(self.errors.get())
+        if len(errs) > 0:
+            raise Exception("Errors: %s" % ' , '.join(errs))
+
+    @staticmethod
+    def _get_test_class_instance(q, err):
+        try:
+            obj = TestClassToTestSingletonPerThread()
+            q.put(obj)
+        except Exception as e:
+            err.put(str(e))
+
+    def _parse_output_and_return_thread_objects(self, queue_output):
+        obj1, obj2 = queue_output.get(), queue_output.get()
+
+        def check_obj(name):
+            if obj1.name == name:
+                return obj1
+            elif obj2.name == name:
+                return obj2
+            else:
+                return None
+
+        t1_object = check_obj(self.THREAD_NAME_1)
+        t2_object = check_obj(self.THREAD_NAME_2)
+
+        return t1_object, t2_object
+
+    def test_it_should_have_only_one_instance_for_same_thread(self):
+        obj1 = TestClassToTestSingletonPerThread()
+        obj2 = TestClassToTestSingletonPerThread()
+
+        self.assertEqual(obj1.uuid, obj2.uuid)
```

Comment:
minor: uuid is ok, but you could also just compare the references:

self.assertEqual(obj1, obj2)

Reply:
ok; maybe worth a comment in the code, since this would be unexpected to the uninitiated

Reply:
Makes sense, I'll add a comment